libs/backend/src/cast_vote_records/referenced_files.ts

```diff
@@ -0,0 +1,123 @@
+}
+
+/**
+ * The files referenced by a cast vote record report
+ */
+export interface ReferencedFiles {
+  imageFiles: SheetOf<ReferencedFile<Buffer>>;
+  layoutFiles?: SheetOf<ReferencedFile<BallotPageLayout>>;
+}
+
+/**
+ * An image referenced by a cast vote record report. When read, the image's hash will be checked
+ * against the hash in the cast vote record report.
+ */
+export class ReferencedImageFile implements ReferencedFile<Buffer> {
+  private readonly expectedFileHash: string;
+  private readonly filePath: string;
+
+  constructor(input: { expectedFileHash: string; filePath: string }) {
+    this.expectedFileHash = input.expectedFileHash;
+    this.filePath = input.filePath;
+  }
+
+  async read(): Promise<Result<Buffer, ReadCastVoteRecordError>> {
+    let fileContents: Buffer;
+    try {
+      fileContents = await fs.readFile(this.filePath);
+    } catch (error) {
+      if (isNonExistentFileOrDirectoryError(error)) {
+        return err({
+          type: 'invalid-cast-vote-record',
+          subType: 'image-not-found',
+        });
+      }
+      return err({
+        type: 'invalid-cast-vote-record',
+        subType: 'image-read-error',
+      });
+    }
+
+    if (sha256(fileContents) !== this.expectedFileHash) {
+      return err({
+        type: 'invalid-cast-vote-record',
+        subType: 'incorrect-image-hash',
+      });
+    }
+
+    return ok(fileContents);
+  }
+}
+
+/**
+ * A layout file referenced by a cast vote record report. When read, the layout file's hash will be
+ * checked against the hash in the cast vote record report.
+ */
+export class ReferencedLayoutFile implements ReferencedFile<BallotPageLayout> {
+  private readonly expectedFileHash: string;
+  private readonly filePath: string;
+
+  constructor(input: { expectedFileHash: string; filePath: string }) {
+    this.expectedFileHash = input.expectedFileHash;
+    this.filePath = input.filePath;
+  }
+
+  async read(): Promise<Result<BallotPageLayout, ReadCastVoteRecordError>> {
+    let fileContents: string;
+    try {
+      fileContents = await fs.readFile(this.filePath, 'utf-8');
+    } catch (error) {
+      if (isNonExistentFileOrDirectoryError(error)) {
+        return err({
+          type: 'invalid-cast-vote-record',
+          subType: 'layout-file-not-found',
+        });
+      }
+      return err({
+        type: 'invalid-cast-vote-record',
+        subType: 'layout-file-read-error',
+      });
+    }
+
+    if (sha256(fileContents) !== this.expectedFileHash) {
+      return err({
+        type: 'invalid-cast-vote-record',
+        subType: 'incorrect-layout-file-hash',
+      });
+    }
+
+    const parseResult = safeParseJson(fileContents, BallotPageLayoutSchema);
+    if (parseResult.isErr()) {
+      return err({
+        type: 'invalid-cast-vote-record',
+        subType: 'layout-file-parse-error',
+      });
+    }
+    const layout = parseResult.ok();
+
+    return ok(layout);
+  }
+}
```

Comment:
These two classes are really similar. Consider consolidating some of the shared logic in composible functions (as opposed to subclassing)?

Reply:
Updated the code to 1) better share logic and 2) use composable functions (a pattern that I haven't really used before but like!): https://github.com/votingworks/vxsuite/pull/4056/commits/89734d1adb6d1aa39c14280beb164019167c72ed